Common/GPU/Vulkan/VulkanRenderManager.cpp

```diff
@@ -1625,10 +1645,153 @@ VKRPipelineLayout *VulkanRenderManager::CreatePipelineLayout(BindingType *bindin
 
 	vulkan_->SetDebugName(layout->descriptorSetLayout, VK_OBJECT_TYPE_DESCRIPTOR_SET_LAYOUT, tag);
 	vulkan_->SetDebugName(layout->pipelineLayout, VK_OBJECT_TYPE_PIPELINE_LAYOUT, tag);
+
+	for (int i = 0; i < VulkanContext::MAX_INFLIGHT_FRAMES; i++) {
+		layout->frameData[i].pool.Create(vulkan_, bindingTypes, (uint32_t)bindingTypesCount, 512);
+		layout->frameData[i].pool.Setup([]() {});
+	}
+
+	pipelineLayouts_.push_back(layout);
 	return layout;
 }
 
 void VulkanRenderManager::DestroyPipelineLayout(VKRPipelineLayout *layout) {
+	for (int i = 0; i < VulkanContext::MAX_INFLIGHT_FRAMES; i++) {
+		layout->frameData[i].pool.Destroy();
+	}
+
 	vulkan_->Delete().QueueDeletePipelineLayout(layout->pipelineLayout);
 	vulkan_->Delete().QueueDeleteDescriptorSetLayout(layout->descriptorSetLayout);
+	for (auto iter = pipelineLayouts_.begin(); iter != pipelineLayouts_.end(); iter++) {
+		if (*iter == layout) {
+			pipelineLayouts_.erase(iter);
+			break;
+		}
+	}
+}
+
+void VulkanRenderManager::FlushDescriptors(int frame) {
+	for (auto iter : pipelineLayouts_) {
+		iter->FlushDescSets(vulkan_, frame, &frameData_[frame].profile);
+	}
+}
+
+void VulkanRenderManager::ResetDescriptorLists(int frame) {
+	for (auto iter : pipelineLayouts_) {
+		VKRPipelineLayout::FrameData &data = iter->frameData[frame];
+
+		data.flushedDescriptors_ = 0;
+		data.descSets_.clear();
+		data.descData_.clear();
+	}
+}
+
+void VKRPipelineLayout::FlushDescSets(VulkanContext *vulkan, int frame, QueueProfileContext *profile) {
+	_dbg_assert_(frame < VulkanContext::MAX_INFLIGHT_FRAMES);
+
+	FrameData &data = frameData[frame];
+
+	VulkanDescSetPool &pool = data.pool;
+	FastVec<PackedDescriptor> &descData = data.descData_;
+	FastVec<PendingDescSet> &descSets = data.descSets_;
+
+	pool.Reset();
+
+	// This will write all descriptors.
+	// Initially, we just do a simple look-back comparing to the previous descriptor to avoid sequential dupes.
+
+	// Initially, let's do naive single desc set writes.
+	VkWriteDescriptorSet writes[MAX_DESC_SET_BINDINGS];
+	VkDescriptorImageInfo imageInfo[MAX_DESC_SET_BINDINGS];  // just picked a practical number
+	VkDescriptorBufferInfo bufferInfo[MAX_DESC_SET_BINDINGS];
+
+	size_t start = data.flushedDescriptors_;
+	int writeCount = 0;
+
+	for (size_t index = start; index < descSets.size(); index++) {
+		auto &d = descSets[index];
+
+		// This is where we look up to see if we already have an identical descriptor previously in the array.
+		// We could do a simple custom hash map here that doesn't handle collisions, since false positives aren't too bad.
+		// Instead, for now we just check history one item backwards. Good enough, it seems.
+		if (index > start + 1) {
+			if (descSets[index - 1].count == d.count) {
+				if (!memcmp(descData.data() + d.offset, descData.data() + descSets[index - 1].offset, d.count * sizeof(PackedDescriptor))) {
+					d.set = descSets[index - 1].set;
+					continue;
+				}
+			}
+		}
+
+		// TODO: Allocate in batches.
```

Comment:
How bad is it to overallocate?  It'd probably be relatively easy to allocate just in chunks of say 8 or something, exhaust and refill.  Or could just estimate a fraction of `descSets.size()` at the start, and fall back to one-at-a-time if it's somehow over the limit.

Wonder if there's benefit or downside to putting more distance between the allocation and writing (i.e. do we want all writes for the small batches to happen right after the alloc, or better to allocate everything, then do writes in batched chunks...)  Well, just pondering, I'm sure the answer is virtually unknowable with all the drivers out there...

-[Unknown]

Reply:
Yeah, small batch allocations might be the way to go just to reduce function call overhead to basically nothing.

Another variant advocated by ARM for cases like this where we start over every frame, is to simply don't reset and allocate, just stomp over your already allocated descriptor sets.

Many things to try, but now that this is no longer on the main thread, it just doesn't seem to matter hugely and as you say it's unknowable - so I think just doing small block allocations here is the safest way. I'm thinking 16.